articles/po3rin.re

```diff
@@ -0,0 +1,643 @@
+= Go + Gonum による行列計算のエッセンス
+
+== はじめに
+
+白ヤギコーポレーションでバックエンドエンジニアをしている@po3rin@<fn>{po3rin}です。仕事ではGoをメインにサーバーサイドの開発をしています。Goで行列なんて扱えるのかと思うかもしれませんが、Go製の数値計算パッケージである@<em>{Gonum}の中に@<em>{mat}パッケージという行列計算を行うための機能が提供されています。今回はGo+Gonumによる行列計算のエッセンスを紹介します。
+//footnote[po3rin][@<href>{https://twitter.com/po3rin}]
+
+== Gonumによる行列の生成と内部実装
+
+ここからはGonumのドキュメント@<fn>{godoc}を見ながらだと理解が進みやすいと思います。まずは行列の作り方を見てみましょう。
+
+//footnote[godoc][@<href>{https://godoc.org/gonum.org/v1/gonum/mat}]
+
+//list[gen][行列の生成][go]{
+package main
+
+import (
+    "fmt"
+
+    "gonum.org/v1/gonum/mat"
+)
+
+func main() {
+    A := mat.NewDense(3, 4, []float64{
+        1, 2, 3, 4,
+        5, 6, 7, 8,
+        9, 10, 11, 12,
+    })
+    // |1  2  3  4|
+    // |5  6  7  8|
+    // |9 10 11 12|
+
+    // ...
+}
+//}
+
+@<list>{gen}で@<code>{mat.NewDense}を使って@<code>{*mat.Dense}という行列のデータ構造をもつ構造体が返ってきます。@<code>{mat.Dense}のデータ構造をみてみましょう。
+
+//list[inner][mat.Denseの定義][go]{
+// gonum/mat/dense.go
+
+type Dense struct {
+    mat blas64.General
+
+    capRows, capCols int
+}
+
+func NewDense(r, c int, data []float64) *Dense {
+    if r < 0 || c < 0 {
+        panic("mat: negative dimension")
+    }
+    if data != nil && r*c != len(data) {
+        panic(ErrShape)
+    }
+    if data == nil {
+        data = make([]float64, r*c)
+    }
+    return &Dense{
+        mat: blas64.General{
+            Rows:   r,
+            Cols:   c,
+            Stride: c,
+            Data:   data,
+        },
+        capRows: r,
+        capCols: c,
+    }
+}
+//}
+
+@<code>{mat.NewDense}を見れば分かりやすいですが、Denseの正体は要素の値の一覧である@<code>{[]float64}に行数、列数、ストライドなどの情報を保持しているだけです。また、要素を全部0で初期化したい場合は@<code>{mat.NewDense}コードから@<code>{mat.NewDense}の第三引数に@<code>{nil}を渡せばよいことが分かります。mat.NewDense@<code>{Dense.mat}の型である@<code>{blas64.General}については後ほど解説します。そして@<code>{*mat.Dense}は@<code>{mat.Matrix}というインターフェースを実装しています。
+
+//list[matrix][mat.Matrixインターフェース][go]{
+// Matrix is the basic matrix interface type.
+type Matrix interface {
+    Dims() (r, c int) // 行数と列数を返す
+    At(i, j int) float64 // 指定した要素の値を返す
+    T() Matrix // 転置行列を返す
+}
+//}
+
+@<code>{mat.Dense}以外で@<code>{Matrix}インターフェースを実装するものとしては@<code>{mat.VecDense}（ベクトルを表現するデータ構造）や@<code>{mat.TriDense}（三角行列を表現するデータ構造）などさまざまな構造体があります。@<code>{Matrix}インターフェースとして振る舞いを定義することで、データ構造の違いを意識せずにコード内で利用できるようになっています。
+
+== Gonumで簡単な行列計算をやってみよう
+
+=== 行列のPrint
+
+早速Gonumによる行列計算の例をみていきたいところですが。今後の行列のデバッグのために@<list>{print}のように@<code>{mat.Formatted}関数を使って行列の構造を確認できるようにしておくと便利です。@<code>{mat.Formatted}関数は@<em>{Functional Option Pattern}で実装されており、すでに用意されているオプション関数を使って出力する行列をフォーマットできます。Functional Option PatternとはGoでよく使われるデザインパターンの１つなので、気になる方はRob Pike氏の「Self-referential functions and the design of options」@<fn>{fop}という記事を読んでみましょう。
```

Comment:
**[textlint]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub>
Line 88 sentence length(136) exceeds the maximum sentence length of 100.
Over 36 characters. (ja-technical-writing/sentence-length)